R/driver-snowflake.R

```diff
@@ -392,3 +524,358 @@ workbench_snowflake_token <- function(account, sf_home) {
+      ),
+      call = quote(DBI::dbConnect())
+    )
+  }
+
+  # WARNING if readable by others
+  if (other_read) {
+    cli::cli_warn(c(
+      sprintf("Bad owner or permissions on %s.", file_path),
+      "i" = sprintf("To change owner, run `chown $USER \"%s\"`.", file_path),
+      "i" = sprintf("To restrict permissions, run `chmod 0600 \"%s\"`.", file_path),
+      "i" = "To skip this warning, set environment variable",
+      " " = "SF_SKIP_WARNING_FOR_READ_PERMISSIONS_ON_CONFIG_FILE=true."
+    ))
+  }
+
+  invisible()
+}
+
+#' Parse a TOML file with proper error handling
+#'
+#' @param file_path Path to TOML file
+#' @return Named list with parsed TOML content
+#' @noRd
+parse_toml_file <- function(file_path) {
+  check_toml_installed()
+
+  tryCatch(
+    toml::read_toml(file_path),
+    error = function(e) {
+      abort(
+        sprintf("An unknown error happened while loading '%s'", file_path),
+        parent = e,
+        call = quote(DBI::dbConnect())
+      )
+    }
+  )
+}
+
+#' Parse a TOML string (from env var) with proper error handling
+#'
+#' @param toml_string TOML content as a string
+#' @return Named list with parsed TOML content
+#' @noRd
+parse_toml_string <- function(toml_string) {
+  check_toml_installed()
+
+  tryCatch(
+    toml::parse_toml(toml_string),
+    error = function(e) {
+      abort(
+        "Failed to parse SNOWFLAKE_CONNECTIONS environment variable as TOML",
+        parent = e,
+        call = quote(DBI::dbConnect())
+      )
+    }
+  )
+}
+
+#' Check if toml package is installed
+#' @noRd
+check_toml_installed <- function() {
+  if (!rlang::is_installed("toml")) {
+    abort(
+      c(
+        "The {.pkg toml} package is required to load Snowflake configuration files.",
+        "i" = "Install it with: install.packages(\"toml\")"
+      ),
+      call = quote(DBI::dbConnect())
+    )
+  }
+}
+
+#' Load and merge Snowflake configuration from files and environment variables
+#'
+#' @param config_dir Path to Snowflake config directory
+#' @param connections_file_path Optional custom path to connections.toml
+#' @param require_toml If TRUE, error if toml package not installed when files exist
+#' @return Named list with 'default_connection_name' and 'connections'
+#' @noRd
+load_snowflake_config <- function(config_dir, connections_file_path = NULL,
+                                  require_toml = FALSE) {
+  result <- list(
+    default_connection_name = "default",
+    connections = list()
+  )
+
+  # Priority 1: Check SNOWFLAKE_CONNECTIONS env var (highest precedence)
+  env_connections <- Sys.getenv("SNOWFLAKE_CONNECTIONS", "")
+  if (nchar(env_connections) > 0) {
+    result$connections <- parse_toml_string(env_connections)
+    # Also check for default connection name override
+    env_default <- Sys.getenv("SNOWFLAKE_DEFAULT_CONNECTION_NAME", "")
+    if (nchar(env_default) > 0) {
+      result$default_connection_name <- env_default
+    }
+    return(result)
+  }
```

Comment:
If `$SNOWFLAKE_CONNECTIONS`, then config.toml is no longer read--is that the same as Python? Like, could you use `$SNOWFLAKE_CONNECTIONS` to define the connections but config.toml to declare which one should be the default? (It does sound far-fetched and I think either answer would be defensible if the Python reference implementation didn't exist... and arguably still defensible regardless of what Python does.)

Reply:
Actually this function needs to be completely rewritten--as it stands, it's too hard to reason about edge cases.

Reply:
Rewritten to use promises (the other kind of promises), way way simpler to reason about now.